File: leetcode/two pointer/438. Find All Anagrams in a String/js/main.js
// Time O(N^2)
// Space O(N)
const findAnagrams2 = (s, p) => {
  const ans = [];
  const map = {};
  const set = new Set();

  for (let x of p) {
    map[x] = ~~map[x] + 1;
  }

  for (let i = 0; i < s.length; i++) {
    let cnt = 0;
    const map2 = {};
    const str = s.substring(i, i + p.length);

    if (set.has(str)) {
      ans.push(i);
      continue;
    }

    for (let x of str) {
      map2[x] = ~~map2[x] + 1;
      if (map2[x] === map[x]) {
        cnt++;
      }
    }

    if (cnt === Object.keys(map).length) {
      set.add(str);
      ans.push(i);
    }
  }
  return ans;
};

File: leetcode/two pointer/438. Find All Anagrams in a String/py/main.py
import collections


class Solution:
    # Time O(S)
    # Space (P)
    def findAnagrams(self, s, p):
        count = collections.Counter(p)

        start, end = 0, 0
        k = len(count)
        ans = []

        while end < len(s):
            if s[end] in count:
                if count[s[end]] == 1:
                    k -= 1
                count[s[end]] -= 1
            end += 1

            while k == 0:
                if end - start == len(p):
                    ans.append(start)

                if s[start] in count:
                    if count[s[start]] == 0:
                        k += 1
                    count[s[start]] += 1
                start += 1

        return ans